lib/bap/bap_self.ml

```diff
@@ -60,4 +60,66 @@ module Create() = struct
   let warning f = message Warning ~section:name f
   let error f = message Error ~section:name f
 
+  module Config = struct
+    let plugin_name = name
+    include Bap_config
+
+    let (/) = Filename.concat
+    let confdir = confdir / plugin_name
+    let conf_filename = confdir / "config"
+
+    let get_env_options () =
+      let prefix = "BAP_" ^ String.uppercase plugin_name ^ "_" in
+      let prefix_chop_key (k, v) =
+        match String.chop_prefix ~prefix k with
+        | Some k -> Some (k, v)
+        | None -> None in
+      let plugin_filter_map str =
+        match String.split str ~on:'=' with
+        | k :: vs -> prefix_chop_key (k, String.concat ~sep:"=" vs)
+        | [] -> None in
+      Unix.environment () |>
+      Array.to_list |>
+      List.filter_map ~f:plugin_filter_map
+
+    let conf_file_options () =
+      let string_splitter str =
+        match String.split str ~on:'=' with
+        | k :: vs -> k, String.concat ~sep:"=" vs
+        | [] -> invalid_arg "empty string" in
+      let split_filter = List.map ~f:string_splitter in
+      try
+        In_channel.with_file
+          conf_filename ~f:(fun ch -> In_channel.input_lines ch
+                                      |> split_filter)
+      with Sys_error _ -> []
+
+    let options () =
+      get_env_options () |>
+      List.fold ~init:(conf_file_options ())
+        ~f:(fun o (k, v) -> List.Assoc.add o k v)
+
+    let get = options () |> List.Assoc.find ~equal:String.Caseless.equal
+
+    let set ~name ~data =
+      let name = String.lowercase name in
+      let old_conf = conf_file_options () in
+      let remaining_conf = List.Assoc.remove old_conf name in
+      let new_conf = List.Assoc.add remaining_conf name data in
```

Comment:
The OCaml style for this would be:
```ocaml
let conf = conf_file_options () in
let conf = List.Assoc.remove old_conf name in
let conf = List.Assoc.add remaining_conf name data in
```

You shouldn't introduce new names, instead you should kill outdated versions of data as soon is possible. There are many reasons for this:
1. Having all these names in the scope, makes code harder to understand:
1.1. reader may think that you will later use all versions of the variable, including old one, so that's why you gave    different names;
1.2. reader need to start with the initial hypothesis that each name is associated with values of different genera, e.g., different types, intents and so on, otherwise why you gave them different names. 
2. It is error prone, as it so easy to access an outdated version of the data. 
3. It is inefficient.
3.1. You extend a lifetime of a variable more than needed
3.2. You prohibit an optimizations, where a compiler can notice that the old value is not used anywhere and just update it destructively.